### ISSUE TITLE
RunningAverage: Confusion matrix must have at least one example before it can be computed

### ISSUE THREAD

**Issue by schuhschuh** [pytorch/ignite#2158]
## 🐛 Bug description

I tried to use a `DiceCoefficient` metric based on `ConfusionMatrix` in conjunction with a `ignite.contrib.handlers.tqdm_logger.ProgressBar`. This gives the below `NotComputableError`. When removing the running average metric from the `metrics` argument of the progress bar, I still get this error suggesting it is not related to the progress bar (which is attached to the engine after the metrics have been attached). The error also persists when attaching the `RunningAverage` after the `DiceCoefficient`.

```
File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 701, in run
    return self._internal_run()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 774, in _internal_run
    self._handle_exception(e)
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 469, in _handle_exception
    raise e
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 744, in _internal_run
    time_taken = self._run_once_on_dataset()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 848, in _run_once_on_dataset
    self._handle_exception(e)
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 469, in _handle_exception
    raise e
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 835, in _run_once_on_dataset
    self._fire_event(Events.ITERATION_COMPLETED)
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/engine/engine.py", line 424, in _fire_event
    func(*first, *(event_args + others), **kwargs)
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metric.py", line 340, in completed
    result = self.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/running_average.py", line 97, in compute
    self._value = self._get_src_value()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/running_average.py", line 113, in _get_metric_value
    return self.src.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in compute
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in <listcomp>
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 145, in _get_value_on_cpu
    v = v.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in compute
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in <listcomp>
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 145, in _get_value_on_cpu
    v = v.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in compute
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in <listcomp>
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 145, in _get_value_on_cpu
    v = v.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in compute
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in <listcomp>
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 145, in _get_value_on_cpu
    v = v.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in compute
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 90, in <listcomp>
    materialized = [_get_value_on_cpu(i) for i in self.args]
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metrics_lambda.py", line 145, in _get_value_on_cpu
    v = v.compute()
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/metric.py", line 589, in another_wrapper
    return func(self, *args, **kwargs)
  File "/hf/home/aschuh/tools/conda/envs/istn_dl4/lib/python3.8/site-packages/ignite/metrics/confusion_matrix.py", line 139, in compute
    raise NotComputableError("Confusion matrix must have at least one example before it can be computed.")
ignite.exceptions.NotComputableError: Confusion matrix must have at least one example before it can be computed.
```

## Environment

- PyTorch Version (e.g., 1.4): 1.8.1
- Ignite Version (e.g., 0.3.0): 0.4.5
- OS (e.g., Linux): CentOS 7
- How you installed Ignite (`conda`, `pip`, source): conda
- Python version: 3.8.10
- Any other relevant information:

**Comment by vfdev-5** [comment]
@schuhschuh thanks for reporting that ! @trsvchn could you please check it and inspect what is the problem exactly. Thanks

**Comment by schuhschuh** [comment]
The below script reproduces the error:

```python
from typing import Optional, Tuple

from ignite.contrib.handlers.tqdm_logger import ProgressBar
from ignite.engine import Engine
from ignite.metrics import ConfusionMatrix, DiceCoefficient, RunningAverage

import torch
import torch.backends.cudnn
from torch import Tensor


def as_one_hot_tensor(
    tensor: Tensor,
    num_classes: int,
    ignore_index: Optional[int] = None,
    dtype: Optional[torch.dtype] = None,
) -> Tensor:
    r"""Converts label image to one-hot encoding of multi-class segmentation.

    Adapted from: https://github.com/wolny/pytorch-3dunet

    Args:
        tensor: Input tensor of shape ``(N, 1, ..., X)`` or ``(N, C, ..., X)``.
            When a tensor with ``C == num_classes`` is given, it is converted to the specified
            ``dtype`` but not modified otherwise. Otherwise, the input tensor must contain
            class labels in a single channel.
        num_classes: Number of channels/labels.
        ignore_index: Ignore index to be kept during the expansion. The locations of the index
            value in the GT image is stored in the corresponding locations across all channels so
            that this location can be ignored across all channels later e.g. in Dice computation.
            This argument must be ``None`` if ``tensor`` has ``C == num_channels``.
        dtype: Data type of output tensor. Default is ``torch.float``.

    Returns:
        Output tensor of shape ``(N, C, ..., X)``.

    """
    if dtype is None:
        dtype = torch.float
    if not isinstance(tensor, Tensor):
        raise TypeError("as_one_hot_tensor() 'tensor' must be torch.Tensor")
    if tensor.dim() < 3:
        raise ValueError("as_one_hot_tensor() 'tensor' must have shape (N, C, ..., X)")
    if tensor.shape[1] == num_classes:
        return tensor.to(dtype=dtype)
    elif tensor.shape[1] != 1:
        raise ValueError(
            f"as_one_hot_tensor() 'tensor' must have shape (N, 1|{num_classes}, ..., X)"
        )
    # create result tensor shape (NxCxDxHxW)
    shape = list(tensor.shape)
    shape[1] = num_classes
    # scatter to get the one-hot tensor
    if ignore_index is None:
        return torch.zeros(shape, dtype=dtype).to(tensor.device).scatter_(1, tensor, 1)
    # create ignore_index mask for the result
    mask = tensor.expand(shape) == ignore_index
    # clone the src tensor and zero out ignore_index in the inputs
    inputs = tensor.clone()
    inputs[inputs == ignore_index] = 0
    # scatter to get the one-hot tensor
    result = torch.zeros(shape, dtype=dtype).to(inputs.device).scatter_(1, inputs, 1)
    # bring back the ignore_index in the result
    result[mask] = ignore_index
    return result


def main() -> int:

    batch_size = 1
    num_classes = 2
    spatial_shape = (32, 32)

    y_pred_shape = (batch_size, num_classes) + spatial_shape
    y_shape = (batch_size,) + spatial_shape

    def process_function(engine: Engine, batch: Tensor) -> Tuple[Tensor, Tensor]:
        dtype = torch.int32
        y_pred = torch.empty(y_pred_shape).uniform_()
        y_pred = y_pred.round().type(torch.int64)
        y_pred = as_one_hot_tensor(y_pred, num_classes=2, dtype=dtype)
        y = torch.empty(y_shape).uniform_().round().type(dtype)
        return y_pred, y

    engine = Engine(process_function)

    cm = ConfusionMatrix(2)
    dsc = DiceCoefficient(cm)
    running_avg_dsc = RunningAverage(dsc)

    dsc.attach(engine, "dsc")
    running_avg_dsc.attach(engine, "running_avg_dsc")

    pbar = ProgressBar()
    pbar.attach(engine, ["running_avg_dsc"])

    engine.run(list(range(100)))
    return 0


if __name__ == "__main__":
    main()

```

**Comment by trsvchn** [comment]
@schuhschuh thank you for reporting that!
@vfdev-5 sure, I'll check that!

**Comment by trsvchn** [comment]
@schuhschuh thank you for your patience!

Looks like `RunningAverage` in `v0.4.5` is not able to update `MetricsLambda` and it's dependency metrics, like `ConfusionMatrix` in your case.

https://github.com/pytorch/ignite/blob/fbedab9d2ab99d7160388a4c13418087972a7243/ignite/metrics/metrics_lambda.py#L83-L87

Here is another example that you can try to reproduce this kind of  issue:

```python
# v0.4.5
...
    engine = Engine(process_function)

    acc = Accuracy()
    err = (1.0 - acc) * 100.0
    running_avg_err = RunningAverage(err)
    running_avg_err.attach(engine, "running_avg_err")

    engine.run(list(range(100)))
    return 0
...
```
```python
NotComputableError: Accuracy must have at least one example before it can be computed.
```

Recursive update for `MetricsLambda` was implemented in `0.4.6`, see #2091:

https://github.com/pytorch/ignite/blob/84e51c3b4987c16f6dd27a6abd2f9ff95da3dc6c/ignite/metrics/metrics_lambda.py#L85-L96

As you can see in this case update method will be invoked by `RunningAverage`and this time it will recursively update all dependent metrics.

To solve this issue it is recommended to upgrade `ignite` to version `v0.4.6`. In, addition, please, note that in `v0.4.6` no need to append metric when using with `RunningAverage`, just attach `running_avg`:

```python
# v0.4.6
...
    engine = Engine(process_function)

    cm = ConfusionMatrix(2)
    dsc = DiceCoefficient(cm)
    running_avg_dsc = RunningAverage(dsc)

    running_avg_dsc.attach(engine, "running_avg_dsc")

    pbar = ProgressBar()
    pbar.attach(engine, ["running_avg_dsc"])

    engine.run(list(range(100)))
    return 0
...
```

**Comment by schuhschuh** [comment]
Great, thanks for looking into this. Will update Ignite!

**Comment by aschuh-hf** [comment]
@trsvchn I now had a chance to update ignite to 0.4.6 and try this. I suppose it works, but the issue I am facing now is that I cannot attach the metric which the `RunningAverage` is based on to the engine. As I am using the running average just for progress status reporting, I am still interested to compute the actual epoch-wise metric that it is based on because it's the latter that I am interested in to report as final validation score. How do you suggest doing this without having to compute to separate confusion matrices?

As provided in my example code, I seem to really want and need to attach both metrics:
```
engine = Engine(process_function)

cm = ConfusionMatrix(2)
dsc = DiceCoefficient(cm)  # Actual validation metric of interest, with EpochWise metric usage
running_avg_dsc = RunningAverage(dsc)  # To be computed for progress bar status only

dsc.attach(engine, "dsc")
running_avg_dsc.attach(engine, "running_avg_dsc")

pbar = ProgressBar()
pbar.attach(engine, ["running_avg_dsc"])
```

**Comment by aschuh-hf** [comment]
The work-around I thought of gives me another problem. I have a custom `output_transform` attached to `ConfusionMatrix` which converts the binary `y_pred` engine outputs into one-hot encoded 2-class vectors as required. This works fine when I just attach the `dsc = DiceCoefficient(cm)` metric lambda to the engine. However, when I use `RunningAverage(DiceCoefficient(cm))`, I receive the following error:

```
y_pred does not have correct number of classes: 1 vs 2
```

This appears to be caused by the output transform of `cm = ConfusionMatrix(2, output_transform=...)` not being invoked when I attach the running average metric to the engine as in:
```
RunningAverage(DiceCoefficient(cm)).attach(engine, "running_avg_dsc")
```

**Comment by trsvchn** [comment]
@schuhschuh Thank you for the questions! I'll start with the second one:
> The work-around I thought of gives me another problem. I have a custom `output_transform` attached to `ConfusionMatrix` which converts the binary `y_pred` engine outputs into one-hot encoded 2-class vectors as required. This works fine when I just attach the `dsc = DiceCoefficient(cm)` metric lambda to the engine. However, when I use `RunningAverage(DiceCoefficient(cm))`, I receive the following error:
> ...

by default, `output_transform` is called only in attached mode, (similar issue #2159). it is applied in `iteration_completed`:

https://github.com/pytorch/ignite/blob/0e0200d7911776f5cd378f6ef29fde169836837c/ignite/metrics/running_average.py#L123
then
https://github.com/pytorch/ignite/blob/0e0200d7911776f5cd378f6ef29fde169836837c/ignite/metrics/metric.py#L296
then this recursive call in case of `MetricsLambda`:
https://github.com/pytorch/ignite/blob/0e0200d7911776f5cd378f6ef29fde169836837c/ignite/metrics/metric.py#L322
as you can see `MetricsLambda` doesn't implement its own `iteration_completed` so no propagation, only `update` is recursive but not the whole `iteration_completed` method. so, yeah, thank you for pointing this out! I think this might be a use case for the #2159.

**Comment by trsvchn** [comment]
Agreed! it makes not to compute CM 2 times. But, unfortunately, current api doesn't allow you to use the same attached metric in epochwise and running avg mode. Since, for the first mode we need to accumulate batch results. for the second RunningAvg resets metric before update to ensure that it wasn't no side updates. So, you have to store 2 values one for epochwise compute and one for batchwise compute. Here is what first comes to my mind, maybe it's a bit too much, but anyway, maybe it will be useful:
```python
from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
from ignite.exceptions import NotComputableError


class MyConfusionMatrix(ConfusionMatrix):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # Introduce new flag
        self.running_avg_mode = False

    @reinit__is_reduced
    def reset(self) -> None:
        # Now we reset differently
        if self.running_avg_mode:
            self._confusion_matrix = torch.zeros(self.num_classes, self.num_classes, dtype=torch.int64, device=self._device)
            self.__num_examples = 0
        else:
            self.confusion_matrix = torch.zeros(self.num_classes, self.num_classes, dtype=torch.int64, device=self._device)
            self._num_examples = 0

    @reinit__is_reduced
    def update(self, output):
        self._check_shape(output)
        y_pred, y = output[0].detach(), output[1].detach()

        # Store num_examples for the current batch
        self.__num_examples = y_pred.shape[0]
        # Accumulate for the epoch:
        self._num_examples += self.__num_examples

        # target is (batch_size, ...)
        y_pred = torch.argmax(y_pred, dim=1).flatten()
        y = y.flatten()

        target_mask = (y >= 0) & (y < self.num_classes)
        y = y[target_mask]
        y_pred = y_pred[target_mask]

        indices = self.num_classes * y + y_pred
        m = torch.bincount(indices, minlength=self.num_classes ** 2).reshape(self.num_classes, self.num_classes)
        self.confusion_matrix += m.to(self.confusion_matrix)

        # Store result for the current batch
        self._confusion_matrix = m

    @sync_all_reduce("confusion_matrix", "_num_examples")
    def compute(self) -> torch.Tensor:
        if self.running_avg_mode:
            # Running Avg mode
            cm = self._confusion_matrix.float()
            ne = self.__num_examples
        else:
            # Default mode
            cm = self.confusion_matrix.float()
            ne = self._num_examples

        if ne == 0:
            raise NotComputableError("Confusion matrix must have at least one example before it can be computed.")

        if self.average:
            if self.average == "samples":
                return cm / ne
            else:
                return self.normalize(cm, self.average)

        return cm


def main() -> int:
...
    engine = Engine(process_function)

    cm = MyConfusionMatrix(2)
    dsc = DiceCoefficient(cm)
    dsc.attach(engine, "dsc")  # epochwise

    def compute_run_avg(output):
        # Turn on running_avg mode
        cm.running_avg_mode = True
        # Compute dice score based on self._confusion_matrix
        score = dsc.compute()
        # Reset
        cm.reset()
        # Turn off running_avg mode
        cm.running_avg_mode = False
        return score
    
    hacked_running_avg_dsc = RunningAverage(output_transform=compute_run_avg)
    # Note we do not use src here, just `output_transform` on ouputs of `process_function`!
    hacked_running_avg_dsc.attach(engine, "hacked_running_avg_dsc")

    pbar = ProgressBar()
    pbar.attach(engine, ["hacked_running_avg_dsc"])
```

so, we introduce a flag and computes CM based on that: in epoch   wise manner or running avg, the key thing is also `compute_run_avg` that reused already computed value for the batch, so we compute CM only once.

**Comment by aschuh-hf** [comment]
Thanks for clarifying! That we have to maintain two separate accumulators makes sense. Interesting idea to work around this and avoid having to compute the batch confusion matrix twice. I am thinking this could maybe extended for reusability by let's say the metric which computes the confusion for a given engine output, and then use something akin to `Average` or `RunningAverage` to obtain the epoch-wise average for this tensor-valued metric. Possibly could already use the `ConfusionMatrix` in batch-wise mode for the first part?

**Comment by vfdev-5** [comment]
@schuhschuh I'm a bit curious about the way how you are using metrics, i.e. DiceCoefficient and RunningAverage. I'm assuming that you would like to compute dice coefficient on a validation set using a validation engine ?
In this case, won't it be more interesting to do the following: call explicitly `compute` in parallel on every iteration while reusing the same metric:
```python
import time
import torch
from ignite.engine import Engine, Events
from ignite.metrics import DiceCoefficient, ConfusionMatrix
from ignite.contrib.handlers import ProgressBar

def eval_step(e, b):
    time.sleep(0.5)
    return torch.rand(4, 2), torch.randint(0, 2, size=(4, ))

engine = Engine(eval_step)

cm = ConfusionMatrix(2)
dsc = DiceCoefficient(cm)
dsc.attach(engine, "dsc")

@engine.on(Events.ITERATION_COMPLETED)
def create_latest_dsc():
    engine.state.metrics["latest_dsc"] = dsc.compute()[1]

pbar = ProgressBar(desc="Evaluation")
pbar.attach(engine, metric_names=["latest_dsc", ])

engine.run([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
engine.state.metrics
> Evaluation: [3/10] 30% , latest_dsc=0.2 [00:01<00:03]
```

**Comment by schuhschuh** [comment]
Indeed, this would work for my use case I think. Your suggestion makes me think that maybe I just needed a different MetricsUsage, something in-between BatchWise and EpochWise, where the computation of the metric is triggered both every iteration and at the end of the epoch, but reset is only called at the beginning.

The use case is that I have a standalone validation script which I run with a given model checkpoint (not interleaved with training). The final validation metric is of interest and should be reported. This works fine with Ignite's metrics. I then just thought it would be nice to have a progress bar to see earlier on how things might look in the end without having to wait until the entire dataset is processed. So I tried out the PrograssBar and run into the reported issue. Seemingly due to a misuse / unintended use of the metrics.

**Comment by vfdev-5** [comment]
Indeed, maybe we could enable a combination of metric usages...
Thanks for details on the usage!
